int32 xr_int
int32 xl_int
int32 y_int
int32 zfr_int
int32 zfl_int
int32 zb_int
